lib/src/package_analysis/summary.dart

```diff
@@ -0,0 +1,49 @@
+import 'dart:io';
+
+import 'package:analyzer/dart/analysis/context_builder.dart';
+import 'package:analyzer/dart/analysis/context_locator.dart';
+import 'package:analyzer/dart/analysis/results.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:path/path.dart' as path;
+
+import './common.dart';
+
+Future<void> generateSummary(String packageLocation) async {
+  if (!await FileSystemEntity.isDirectory(packageLocation)) {
+    exitFatal('Please specify a directory for analysis.', 1);
+  }
+
+  // initialise analyser
+  final context = ContextBuilder().createContext(
+    contextRoot: ContextLocator().locateRoots(
+      // use canonicalize to ensure we get an absolute path
+      includedPaths: [path.canonicalize(packageLocation)],
+    ).first,
+  );
+
+  for (final filePath in context.contextRoot.analyzedFiles()) {
+    final session = context.currentSession;
+    final someResolvedLibrary = await session.getResolvedLibrary(filePath);
+
+    // match [packageLocation]/lib/*.dart
+    // but exclude [packageLocation]/lib/src/*.dart
+    if (!(path.isWithin(path.join(packageLocation, 'lib'), filePath) &&
+        !path.isWithin(path.join(packageLocation, 'lib', 'src'), filePath) &&
+        path.extension(filePath) == '.dart')) {
+      continue;
+    }
+
+    // ensure that resolving has been successful
+    if (someResolvedLibrary is! ResolvedLibraryResult) {
+      stderr.writeln('$filePath is not a library');
+      continue;
+    }
+
+    _traverseLibrary(someResolvedLibrary.element);
+  }
+}
+
+void _traverseLibrary(LibraryElement libraryElement) {
+  print(libraryElement.identifier);
+  // TODO
```

Comment:
Also reading that document, I realise I haven't touched a `CompilationUnit` at all in my sample project at all - I obtained a `LibraryElement` instead for every .dart file in the package folder. What's the difference between the two?

It looks like both represent a resolved AST, but a library has to be explicitly defined using the `library` directive (and can be split between different files using `part`). Does every .dart file have its own `CompilationUnit` and vice versa?

It looks like symbols don't have to be part of a library in order to be public and importable, so I should probably work with units instead of libraries. Right?

Reply:
I think each dart file is represented by a `CompilationUnit`, and a LibraryElement that is split in parts is then represented as several `CompilationUnit`s

(I think `CompilationUnit` is kind of a mis-nomer as I would expect the unit of compilation to actually be the library...)